browser/brave_wallet/solana_provider_renderer_browsertest.cc

```diff
@@ -0,0 +1,973 @@
+  }
+
+  void SetEmitEmptyAccountChanged(bool value) {
+    emit_empty_account_changed_ = value;
+  }
+
+ private:
+  void ClearError() {
+    error_ = SolanaProviderError::kSuccess;
+    error_message_.clear();
+  }
+  SolanaProviderError error_ = SolanaProviderError::kSuccess;
+  std::string error_message_;
+  bool emit_empty_account_changed_ = false;
+  mojo::Remote<brave_wallet::mojom::SolanaEventsListener> events_listener_;
+};
+
+class TestBraveContentBrowserClient : public BraveContentBrowserClient {
+ public:
+  TestBraveContentBrowserClient() = default;
+  ~TestBraveContentBrowserClient() override = default;
+  TestBraveContentBrowserClient(const TestBraveContentBrowserClient&) = delete;
+  TestBraveContentBrowserClient& operator=(
+      const TestBraveContentBrowserClient&) = delete;
+
+  void RegisterBrowserInterfaceBindersForFrame(
+      content::RenderFrameHost* render_frame_host,
+      mojo::BinderMapWithContext<content::RenderFrameHost*>* map) override {
+    BraveContentBrowserClient::RegisterBrowserInterfaceBindersForFrame(
+        render_frame_host, map);
+    // override binding for SolanaProvider
+    map->Add<brave_wallet::mojom::SolanaProvider>(
+        base::BindRepeating(&TestBraveContentBrowserClient::BindSolanaProvider,
+                            weak_ptr_factory_.GetWeakPtr()));
+  }
+
+  TestSolanaProvider* GetProvider(content::RenderFrameHost* frame_host) {
+    if (!provider_map_.contains(frame_host))
+      return nullptr;
+    return static_cast<TestSolanaProvider*>(
+        provider_map_.at(frame_host)->impl());
+  }
+  bool WaitForBinding(content::RenderFrameHost* render_frame_host,
+                      base::OnceClosure callback) {
+    if (IsBound(render_frame_host))
+      return false;
+    quit_on_binding_ = std::move(callback);
+    return true;
+  }
+  bool IsBound(content::RenderFrameHost* frame_host) {
+    return provider_map_.contains(frame_host);
+  }
+
+ private:
+  void BindSolanaProvider(
+      content::RenderFrameHost* const frame_host,
+      mojo::PendingReceiver<brave_wallet::mojom::SolanaProvider> receiver) {
+    auto provider = mojo::MakeSelfOwnedReceiver(
+        std::make_unique<TestSolanaProvider>(), std::move(receiver));
+    provider->set_connection_error_handler(
+        base::BindOnce(&TestBraveContentBrowserClient::OnDisconnect,
+                       weak_ptr_factory_.GetWeakPtr(), frame_host));
+    provider_map_[frame_host] = provider;
+    if (quit_on_binding_)
+      std::move(quit_on_binding_).Run();
+  }
+  void OnDisconnect(content::RenderFrameHost* frame_host) {
+    provider_map_.erase(frame_host);
+  }
+
+  base::OnceClosure quit_on_binding_;
+  base::flat_map<
+      content::RenderFrameHost*,
+      mojo::SelfOwnedReceiverRef<brave_wallet::mojom::SolanaProvider>>
+      provider_map_;
+  base::WeakPtrFactory<TestBraveContentBrowserClient> weak_ptr_factory_{this};
+};
+
+}  // namespace
+
+class SolanaProviderRendererTest : public InProcessBrowserTest {
```

Comment:
I believe this test belongs in renderer/brave_wallet

Reply:
yes, I will make it into `renderer/brave_wallet`

Reply:
actually put it under renderer will cause layer violation for header include like `#include "chrome/browser/profiles/profile.h"`